cli/src/cli_util.rs

```diff
@@ -2059,12 +2059,25 @@ See https://jj-vcs.github.io/jj/latest/working-copy/#stale-working-copy \
             .map(|commit_id| tx.repo().store().get_commit(commit_id))
             .transpose()?;
 
+        // Determine if we need to acquire the WC lock:
+        // - If we're exporting Git HEAD (colocated repo with new WC commit)
+        // - OR if we're updating the working copy files
+        let mut locked_ws = {
+            let git_head_export =
+                self.working_copy_shared_with_git && maybe_new_wc_commit.is_some();
+            let wc_update = self.may_update_working_copy && maybe_new_wc_commit.is_some();
+            if git_head_export || wc_update {
+                Some(self.workspace.start_working_copy_mutation()?)
+            } else {
+                None
+            }
+        };
+
         #[cfg(feature = "git")]
         if self.working_copy_shared_with_git {
             use std::error::Error as _;
             if let Some(wc_commit) = &maybe_new_wc_commit {
-                // This can fail if HEAD was updated concurrently. In that case,
-                // the actual state will be imported on the next snapshot.
+                // Export Git HEAD while holding the WC lock to prevent races
                 match jj_lib::git::reset_head(tx.repo_mut(), wc_commit) {
```

Comment:
This is an idea for future change, but I think we can get rid of `view.git_head` and `view.git_refs` by deprecating revset/template functions. These fields track the last known Git state and will never be restored on `jj undo`. Therefore, these fields can be extracted to a separate storage, and `git::reset_head()` can be moved after `tx.commit()`.

I think we have had a failed attempt for this change, but we didn't have `@git` remote at that time.